File: backend/api/schemas/item.py
"""API schemas for the items endpoints.

The DB stores event_date/event_time; the API contract uses date/time
(matching FamilyItem). item_to_read() maps between them.
"""

from typing import Literal

from pydantic import BaseModel


class ItemRead(BaseModel):
    id: str
    item_type: str
    status: str
    event_title: str | None = None
    action_required: str | None = None
    date: str | None = None
    time: str | None = None
    location: str | None = None
    child_name: str | None = None
    event_type: str | None = None
    source_sender: str | None = None
    source_email_link: str | None = None


class ItemListResponse(BaseModel):
    items: list[ItemRead]


class ItemUpdate(BaseModel):
    status: Literal["done", "dismissed"]


def item_to_read(item) -> ItemRead:
    return ItemRead(
        id=str(item.id),
        item_type=item.item_type,
        status=item.status,
        event_title=item.event_title,
        action_required=item.action_required,
        date=item.event_date,
        time=item.event_time,
        location=item.location,
        child_name=item.child_name,
        event_type=item.event_type,
        source_sender=item.source_sender,
        source_email_link=item.source_email_link,
    )

File: frontend/lib/items/item.dart
/// A family item (event or action) as returned by GET /api/v1/items.
/// Mirrors the backend ItemRead schema; `date`/`time` are strings (the backend
/// stores them as-is from extraction).
class Item {
  const Item({
    required this.id,
    required this.itemType,
    required this.status,
    this.eventTitle,
    this.actionRequired,
    this.date,
    this.time,
    this.location,
    this.childName,
    this.eventType,
    this.sourceSender,
    this.sourceEmailLink,
  });

  final String id;
  final String itemType; // "event" | "action"
  final String status; // "open" | "done" | "dismissed"
  final String? eventTitle;
  final String? actionRequired;
  final String? date;
  final String? time;
  final String? location;
  final String? childName;
  final String? eventType;
  final String? sourceSender;
  final String? sourceEmailLink;

  bool get isEvent => itemType == 'event';

  /// Best display label for a list row.
  String get title =>
      eventTitle ?? actionRequired ?? (isEvent ? '(untitled event)' : '(untitled action)');

  /// Defensive variant for list parsing: one malformed row (partial write,
  /// migration hiccup) returns null instead of taking down the whole list —
  /// callers skip nulls.
  static Item? tryParse(Object? json) {
    if (json is! Map) return null;
    try {
      return Item.fromJson(Map<String, dynamic>.from(json));
    } catch (_) {
      return null;
    }
  }

  factory Item.fromJson(Map<String, dynamic> json) => Item(
        id: json['id'] as String,
        itemType: json['item_type'] as String,
        status: json['status'] as String,
        eventTitle: json['event_title'] as String?,
        actionRequired: json['action_required'] as String?,
        date: json['date'] as String?,
        time: json['time'] as String?,
        location: json['location'] as String?,
        childName: json['child_name'] as String?,
        eventType: json['event_type'] as String?,
        sourceSender: json['source_sender'] as String?,
        sourceEmailLink: json['source_email_link'] as String?,
      );
}
